x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Detector.java

```diff
@@ -691,96 +693,36 @@ private static boolean containsInvalidChar(String field) {
-                String msg =
-                        Messages.getMessage(Messages.JOB_CONFIG_DETECTION_RULE_INVALID_TARGET_FIELD_NAME, analysisFields, targetFieldName);
+        private void checkFunctionHasRuleSupport(DetectionRule rule, DetectorFunction function) {
+            if (ruleHasConditionOnResultValue(rule) && FUNCTIONS_WITHOUT_RULE_CONDITION_SUPPORT.contains(function)) {
+                String msg = Messages.getMessage(Messages.JOB_CONFIG_DETECTION_RULE_NOT_SUPPORTED_BY_FUNCTION, function);
                 throw ExceptionsHelper.badRequestException(msg);
             }
         }
 
-        private List<String> getValidFieldNameOptionsForNumeric(DetectionRule rule) {
-            List<String> result = new ArrayList<>();
-            if (overFieldName != null) {
-                result.add(byFieldName == null ? overFieldName : byFieldName);
-            } else if (byFieldName != null) {
-                result.add(byFieldName);
-            }
-
-            if (rule.getTargetFieldName() != null) {
-                ScopingLevel targetLevel = ScopingLevel.from(this, rule.getTargetFieldName());
-                result = result.stream().filter(field -> targetLevel.isHigherThan(ScopingLevel.from(this, field)))
-                        .collect(Collectors.toList());
-            }
-
-            if (isEmptyFieldNameAllowed(rule)) {
-                result.add(null);
-            }
-            return result;
-        }
-
-        private boolean isEmptyFieldNameAllowed(DetectionRule rule) {
-            List<String> analysisFields = extractAnalysisFields();
-            return analysisFields.isEmpty() || (rule.getTargetFieldName() != null && analysisFields.size() == 1);
-        }
-
-        enum ScopingLevel {
-            PARTITION(3),
-            OVER(2),
-            BY(1);
-
-            int level;
-
-            ScopingLevel(int level) {
-                this.level = level;
-            }
-
-            boolean isHigherThan(ScopingLevel other) {
-                return level > other.level;
-            }
-
-            static ScopingLevel from(Detector.Builder detector, String fieldName) {
-                if (fieldName.equals(detector.partitionFieldName)) {
-                    return ScopingLevel.PARTITION;
-                }
-                if (fieldName.equals(detector.overFieldName)) {
-                    return ScopingLevel.OVER;
-                }
-                if (fieldName.equals(detector.byFieldName)) {
-                    return ScopingLevel.BY;
+        private static boolean ruleHasConditionOnResultValue(DetectionRule rule) {
+            for (RuleCondition condition : rule.getConditions()) {
+                switch (condition.getAppliesTo()) {
+                    case ACTUAL:
+                    case TYPICAL:
+                    case DIFF_FROM_TYPICAL:
+                        return true;
+                    case TIME:
+                    default:
```

Comment:
Should `default` throw an exception rather than return false?  If a future enum value is added we don't know what would be most appropriate to return here.